### ISSUE TITLE
Port the portals and rooms occlusion culling (and mesh merging) to Godot 4

### ISSUE THREAD

**Issue by Calinou** [godotengine/godot-proposals#3920]
### Describe the project you are working on

The Godot editor :slightly_smiling_face: 

### Describe the problem or limitation you are having in your project

Godot 4.0 features raster occlusion culling. This occlusion culling method is easy to set up, but it has a few downsides:

- Gameplay notifications cannot be used, as there is no concept of secondary visibility with raster occlusion. This means that AI, physics or complex processing cannot be automatically disabled in areas that are occluded. (It may be possible to allow VisibilityNotifier to check for this, which would allow doing something similar to gameplay notifications with raster occlusion.)
- If the occluder meshes are complex, rendering the scene on the CPU for raster occlusion can take up a significant amount of CPU time. This can make reaching high framerates difficult, regardless of the GPU used. This can be especially the case on mobile.
  - In contrast, portals and rooms occlusion is cheaper on the CPU, even in complex scenes.

In addition to occlusion culling, the portals and rooms system features optional mesh merging to merge several meshes within a room. This can be used to reduce the number of draw calls to further improve performance. While draw calls are cheaper with Vulkan compared to OpenGL, they're not free either.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Port the [portals and rooms](https://docs.godotengine.org/en/3.4/tutorials/3d/portals/index.html) occlusion culling system to Godot 4.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

This port is likely not trivial, as parts of the renderer have been refactored to make culling faster and more parallel when possible.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

No, as a complete, user-friendly portals and rooms system requires a lot of code and editor tooling.

### Is there a reason why this should be core and not an add-on in the asset library?

Occlusion culling needs to be implemented in core to be fast, as it needs to interact with low-level rendering culling classes.

**Comment by tinmanjuggernaut** [comment]
When mesh merging gets implemented in 4, please note #61575. 

Materials should be returned where they were placed by gamedevs, and should not be moved anywhere else or reordered. Each material property has their own purpose. All materials at every level should be copied to the merged mesh:

Master mesh materials are imported here: `Mesh/surface_X/material`
Per Instance, per surface overrides: `MeshInstance3D/surface_material_override/X`
All surface override: `GeometryInstance3D/material_override`

**Comment by Capewearer** [comment]
Could you somehow overcome CPU spikes in large scenes? Is it possible to have several baked occluders which are culled with other occluders or culled by range of camera?

**Comment by Calinou** [comment]
> or culled by range of camera?

Yes, you can use [visibility ranges](https://docs.godotengine.org/en/latest/tutorials/3d/visibility_ranges.html) to hide certain objects when they're far away from the camera.

**Comment by Capewearer** [comment]
Maybe that should be written in documentation? I mean in this article. https://docs.godotengine.org/en/stable/tutorials/3d/occlusion_culling.html

**Comment by Calinou** [comment]
> Maybe that should be written in documentation? I mean in this article. [docs.godotengine.org/en/stable/tutorials/3d/occlusion_culling.html](https://docs.godotengine.org/en/stable/tutorials/3d/occlusion_culling.html)

It's already mentioned at the bottom of [this section](https://docs.godotengine.org/en/stable/tutorials/3d/occlusion_culling.html#why-use-occlusion-culling) :slightly_smiling_face:

**Comment by Capewearer** [comment]
Sorry, but this sentence is too abstract to expliciitly claim the usage of _different_ OccluderInstance3D nodes with switching visibilty of such nodes with script.

**Comment by seacat17** [comment]
What if I'm using Compatibility pipeline that does not support all those nifty-lifty performance enhancements? I need to squeeze out every single frame out of my project, so I need PVS. Rooms and portals.

**Comment by Calinou** [comment]
> What if I'm using Compatibility pipeline that does not support all those nifty-lifty performance enhancements? I need to squeeze out every single frame out of my project, so I need PVS. Rooms and portals.

The Compatibility renderer supports automatic mesh LOD, visibility ranges *and* occlusion culling, just like Forward+/Mobile. This is the case as far back as 4.0.

**Comment by seacat17** [comment]
> > What if I'm using Compatibility pipeline that does not support all those nifty-lifty performance enhancements? I need to squeeze out every single frame out of my project, so I need PVS. Rooms and portals.
> 
> The Compatibility renderer supports automatic mesh LOD, visibility ranges *and* occlusion culling, just like Forward+/Mobile. This is the case as far back as 4.0.

And for some reason in Compatibility mode I need to have that "Render Opacity Pass" thingy in my frametime that eats 70% of performance on lowest-end systems, despite the scene looking like Quake from 1996. I can get rid of that by disabling all the prepasses in the mesh settings, but then the obvious issue arises - the mesh is visible when it shouldn't be.

In the latest Godot release occlusion culling CANNOT eliminate this issue ~~(occlusion culling is just bad in general if you ask me, has so much problems for benefits only in specific cases)~~, so if I want to cull the things the old-fashioned way I need to code the visibility logic by hand which is a dumb effort considering the fact that Godot already supported PVS before and it can support it again. The scene is purely indoors with almost no areas intersecting with the outer space and even these can be eliminated if needed.

**Comment by Calinou** [comment]
> And for some reason in Compatibility mode I need to have that "Render Opacity Pass" thingy in my frametime that eats 70% of performance on lowest-end systems, despite the scene looking like Quake from 1996. I can get rid of that by disabling all the prepasses in the mesh settings, but then the obvious issue arises - the mesh is visible when it shouldn't be.

In the visual profiler, **Render Opaque Pass** refers to the cost of rendering opaque objects in the scene. It's expected to be in the 70%+ range in simple projects, especially if you have little to no post-processing effects enabled.

If you set objects to be rendered in the transparent pass (which happens if you change **Depth Prepass** away from the default), the cost will be removed from this pass, but it'll be added to the transparent pass where it's much slower.

I would also look at the draw call count and make sure it's not excessive due to the level being split up in too many parts (e.g. TrenchBroom imports with one draw call per brush).

> considering the fact that Godot already supported PVS before and it can support it again. The

Unfortunately, the portals and rooms system proved to be scarcely used in Godot 3. It had poor usability and despite all the documentation written for it, a lot of people weren't able to get it working in the first place in their projects.

**Comment by seacat17** [comment]
> a lot of people weren't able to get it working in the first place in their projects.

Was it because of leaks?

> If you set objects to be rendered in the transparent pass (which happens if you change Depth Prepass away from the default), the cost will be removed from this pass, but it'll be added to the transparent pass where it's much slower.

What if I set it to "none"?

**Comment by Calinou** [comment]
> What if I set it to "none"?

The object isn't writing to depth in this case, and is therefore going through the transparent pass. See https://github.com/godotengine/godot-proposals/issues/12923 for a list of reasons that cause a material to be rendered in the transparent pass.
